### PR TITLE
support postgresql 15

### PR DESCRIPTION
**Pull request by Leo-XM-Zeng**

### CODE REVIEW

**Comment by Leo-XM-Zeng** [comment]
image

**But there's a little problem I don't know how to deal with**
image
**It cannot be overwritten directly here for fear of affecting versions higher than PG15**
image

**Comment by Leo-XM-Zeng** [comment]
'\getenv pwd PWD' is better transformed into '\set pwd' pwd ', which is more general and suitable for more PG versions. #178

**Comment by Leo-XM-Zeng** [comment]
> Ragarding the problem, you need to apply the equivalent of this change to the ruleutils.c file too: [d62ba97](https://github.com/duckdb/pg_duckdb/commit/d62ba97b5fe6f1628493ae7aba087cf9621b0872)

That's one step missing. Thank you for your help
image

**Comment by Leo-XM-Zeng** [comment]
@JelteF Hi JelteF, I have a question and I hope you can help me. I wonder if I need to move similar code to "pg_ruleutils_xx.c"? But it doesn't actually belong to "pg_ruleutils". 6025369b695bf743dec83773be3e2f5eb59776e1
image

**Comment by Leo-XM-Zeng** [comment]
Or is there something that needs to be improved in this pull request and is looking forward to your suggestions

File: src/utility/copy.cpp
Comment (JelteF) [inline src/utility/copy.cpp:168]: please install editorconfig and save this file again. We want a final newline in the file. https://editorconfig.org/

Comment (Leo-XM-Zeng) [reply src/utility/copy.cpp]: ok

**Comment by JelteF** [comment]
Can you make sure this is also run in CI, by adding `REL_15_STABLE` here: https://github.com/duckdb/pg_duckdb/blob/ee3526d4f065f1f756caea80758653abfac1eda8/.github/workflows/build_and_test.yaml#L57

File: src/vendor/pg_ruleutils_15.c
Comment (JelteF) [inline src/vendor/pg_ruleutils_15.c:7]: I think this date change was probably done by accident.

Comment (Leo-XM-Zeng) [reply src/vendor/pg_ruleutils_15.c:7]: > I think this date change was probably done by accident. Can you revert that?

Ok, I'll be more careful next time

**Comment by Leo-XM-Zeng** [comment]
> Can you make sure this is also run in CI, by adding here: `REL_15_STABLE`
> 
> https://github.com/duckdb/pg_duckdb/blob/ee3526d4f065f1f756caea80758653abfac1eda8/.github/workflows/build_and_test.yaml#L57

Okay, thanks. I'll try.

File: src/utility/copy.cpp
Comment (JelteF) [inline src/utility/copy.cpp:76]: Does this work? That should allow us to remove the `ListCell` declaration at the top of the function. If not we should update these macros to work for PG15.

```suggestion
	foreach_int(cur, attnums)
	{
		int	attno = cur - FirstLowInvalidHeapAttributeNumber;
```

Comment (Leo-XM-Zeng) [reply src/utility/copy.cpp:76]: This part refers to part of the source code of PG15
image

Comment (JelteF) [reply src/utility/copy.cpp:76]: Hmm, I didn't realize that we copied a part of this code initially. I might have kept the `foreach` macro then when I originally replaced it. But let's stay consistent here for now and use the new `foreach_int` macro in both places.

Comment (Leo-XM-Zeng) [reply src/utility/copy.cpp:76]: I have just read the relevant code and understood your meaning. Try to use the relevant definition of "pg_list.hpp" in the project, and I will modify it. 
```c
#define foreach_ptr(type, var, lst) foreach_internal(type, *, var, lst, lfirst)
#define foreach_int(var, lst)	foreach_internal(int, , var, lst, lfirst_int)
#define foreach_oid(var, lst)	foreach_internal(Oid, , var, lst, lfirst_oid)
#define foreach_xid(var, lst)	foreach_internal(TransactionId, , var, lst, lfirst_xid)

/*
 * The internal implementation of the above macros.  Do not use directly.
 *
 * This macro actually generates two loops in order to declare two variables of
 * different types.  The outer loop only iterates once, so we expect optimizing
 * compilers will unroll it, thereby optimizing it away.
 */
#define foreach_internal(type, pointer, var, lst, func) \
	for (type pointer var = 0, pointer var##__outerloop = (type pointer) 1; \
		 var##__outerloop; \
		 var##__outerloop = 0) \
		for (ForEachState var##__state = {(lst), 0}; \
			 (var##__state.l != NIL && \
			  var##__state.i < var##__state.l->length && \
			 (var = func(&var##__state.l->elements[var##__state.i]), true)); \
			 var##__state.i++)
```

Comment (Leo-XM-Zeng) [reply src/utility/copy.cpp]: Sorry, using part of PG code has become a habit, I did not understand your meaning quickly

File: src/scan/postgres_scan.cpp
Comment (JelteF) [inline src/scan/postgres_scan.cpp:204]: I don't think this is needed for PG15 support (probably this was for PG14 support). So let's remove it for now. Also it's the wrong type afaict. The Cardinality is a typedef for `double`.

File: src/utility/copy.cpp
Comment (JelteF) [inline src/utility/copy.cpp:138]: I guess this change is only needed for PG14 support. Let's remove it from this PR for now.

File: src/pgduckdb_hooks.cpp
Comment (JelteF) [inline src/pgduckdb_hooks.cpp:56]: nitpick: This needs another indent (same for block below), to make it clear that it's part of the `if`

File: src/pgduckdb_hooks.cpp
Comment (JelteF) [inline src/pgduckdb_hooks.cpp:82]: nitpick: this has some trailing whitespace space. Now that you have editorconfig installed, saving this file again should fix that automatically.
```suggestion
#endif
```

File: src/pgduckdb_hooks.cpp
Comment (JelteF) [inline src/pgduckdb_hooks.cpp:58]: Side-note: We might want to create a helper function with a consistent calling convention around such functions, so that we don't need so many `#if PG_VERSION_NUM` statements throughout the codebase. I expect query_tree_walker could become one of those. For now I think this is fine though, since we only call it twice.

But to be clear, let's not do that in this PR.

**Review by JelteF — CHANGES_REQUESTED** [review]
Some final suggestions, but after that it's good to merge IMO.

@wuputah @mkaruza do any of you have some feedback on this. If not I'll squash some of the commits after my final feedback is addressed, and then merge this in a way that preserves the history nicely (specifically the initial copy of the PG15 ruleutils).

File: src/utility/copy.cpp
Comment (Leo-XM-Zeng) [reply src/utility/copy.cpp:138]: Ok, I will submit this paragraph again at PG14

File: src/scan/postgres_scan.cpp
Comment (Leo-XM-Zeng) [reply src/scan/postgres_scan.cpp:204]: Ok, here the type definition basis function "CreateFunctionIndexScanArguments" parameter
image

File: src/pgduckdb_hooks.cpp
Comment (Leo-XM-Zeng) [reply src/pgduckdb_hooks.cpp:82]: ok

File: src/pgduckdb_hooks.cpp
Comment (Leo-XM-Zeng) [reply src/pgduckdb_hooks.cpp:58]: Sorry, I didn't think of a good implementation here, can you help me think of a more elegant implementation to replace this

File: src/scan/postgres_scan.cpp
Comment (JelteF) [reply src/scan/postgres_scan.cpp:204]: Hmm, I guess Postgres and DuckDB cardinalities have a different type (uint64 for Duckdb and double for Postgres). Let's leave it as is for now since this only applies to PG14, but let's add a comment explaining this difference on the PG14 PR.

File: src/pgduckdb_hooks.cpp
Comment (Leo-XM-Zeng) [reply src/pgduckdb_hooks.cpp]: Thank you for your guidance.

**Comment by wuputah** [comment]
> do any of you have some feedback on this ...

Go for it. Tests pass so 👍

**Comment by JelteF** [comment]
@Leo-XM-Zeng sorry for the delay. We had some discussion in person about PG version support, given this is a young project and would like to move fast. The result of that discussion can be fonud here: https://github.com/duckdb/pg_duckdb/issues/154#issuecomment-2382890707

tl;dr we'd like to merge this PG15 support.

Could you fix the merge conflict? Then I'll merge it after.

**Review by JelteF — APPROVED** [review]

**Comment by Leo-XM-Zeng** [comment]
> @Leo-XM-Zeng sorry for the delay. We had some discussion in person about PG version support, given this is a young project and would like to move fast. The result of that discussion can be fonud here: [#154 (comment)](https://github.com/duckdb/pg_duckdb/issues/154#issuecomment-2382890707)
> 
> tl;dr we'd like to merge this PG15 support.
> 
> Could you fix the merge conflict? Then I'll merge it after.
> 
> Thank you for all your work on this.

Ok, just a moment, please.

**Comment by Leo-XM-Zeng** [comment]
@JelteF  oh, this error is due to unsupported syntax in the current psql15 version
image

**Comment by JelteF** [comment]
to be clear, not remove the whole basic.sql just the lines that use `\bind`

**Comment by Leo-XM-Zeng** [comment]
> to be clear, not remove the whole basic.sql just the lines that use `\bind`

understand

**Comment by JelteF** [comment]
@wuputah or @Y-- Could one of you approve this. I've now cleaned up the commits, and I'll "merge" this by pushing straight to master to keep `git blame` working well (just like I did when previously vendoring in the PG16 and PG17 ruleutils files).

**Review by Y-- — APPROVED** [review]

**Pull request merged**
